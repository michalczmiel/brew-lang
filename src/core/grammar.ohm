BrewLang {
  recipe = (line | comment | newline)*
  line = (title | brewer | step | temperature | dose | filter) " "* comment? newline?

  title = "#" " "+ (~(newline | comment) any)+
  brewer = "brewer" " "+ (~(newline | comment) any)+
  filter = "filter" " "+ (~(newline | comment) any)+
  dose = "dose" " "+ number
  temperature = "temperature" " "+ (range | number)+

  step = "at" " " duration_number " "* comment? newline? instruction* "end"
  instruction = " "* (duration | pour | temperature | swirl | stir | comment) (" " | newline | &"end")
  duration = "duration" " " duration_number
  pour = "pour" " " number
  swirl = "swirl"
  stir = "stir"

  comment = "-" "-" (~newline any)*
  newline = "\n"

  range = whole_number "." "." whole_number

  number = real_number | whole_number
  whole_number = digit+
  real_number = digit+ "." digit+

  duration_number = whole_number ":" whole_number
}
