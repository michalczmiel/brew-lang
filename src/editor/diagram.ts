import {
  type Duration,
  type RecipeAST,
  calculateRatioFromAST,
} from "../core/semantics.js";

export interface DiagramOptions {
  width?: number;
  height?: number;
  padding?: number;
}

function formatDuration(duration: Duration): string {
  const minutes = duration.minutes.toString().padStart(1, "0");
  const seconds = duration.seconds.toString().padStart(2, "0");
  return `${minutes}:${seconds}`;
}

function formatTimeRange(start: Duration, end: Duration): string {
  return `${formatDuration(start)} - ${formatDuration(end)}`;
}

function addDuration(base: Duration, toAdd: Duration): Duration {
  const totalSeconds =
    base.minutes * 60 + base.seconds + toAdd.minutes * 60 + toAdd.seconds;
  return { minutes: Math.floor(totalSeconds / 60), seconds: totalSeconds % 60 };
}

const icons = {
  water: `<path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>`,
  coffee: `<circle cx="12" cy="12" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="12" r="1" fill="currentColor"/>`,
  swirl: `<path d="M12 6c0 0 3 2 3 6s-3 6-3 6" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/><path d="M12 6c0 0 -3 2 -3 6s3 6 3 6" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/>`,
  stir: `<circle cx="12" cy="12" r="5" fill="none" stroke="currentColor" stroke-width="1.5" stroke-dasharray="2,2"/><path d="M12 7 L12 17" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/>`,
  thermometer: `<rect x="10" y="4" width="4" height="12" rx="2" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="1.5" fill="currentColor"/><line x1="12" y1="8" x2="12" y2="16" stroke="currentColor" stroke-width="1.5"/>`,
  brewer: `<path d="M8 4h8v3c0 4-2 7-4 7s-4-3-4-7V4z" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M6 4h12" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/><path d="M10 14v4h4v-4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>`,
  timer: `<circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/>`,
} as const;

export function generateSVGDiagram(
  ast: RecipeAST,
  options: DiagramOptions = {},
): string {
  const { width = 600, padding = 20 } = options;

  const contentWidth = width - padding * 2;
  const stepHeight = 80;
  const headerHeight = 120;

  const totalSteps = ast.steps.length;
  const commentsHeight = ast.comments ? ast.comments.length * 20 + 40 : 0;
  const contentHeight =
    headerHeight + totalSteps * stepHeight + 40 + commentsHeight;
  const height = contentHeight + padding * 2;

  let svg = `<svg viewBox="0 0 ${width} ${height}" xmlns="http://www.w3.org/2000/svg" style="font-family: system-ui, -apple-system, sans-serif; background: white;">`;

  svg += `<defs>
    <style>
      .header-text { font-size: 14px; fill: #666; }
      .step-number { font-size: 16px; font-weight: 600; fill: #333; }
      .time-text { font-size: 13px; fill: #666; }
      .instruction-text { font-size: 12px; fill: #444; }
      .icon { color: #333; }
      .divider { stroke: #e0e0e0; stroke-width: 1; stroke-dasharray: 3,3; }
    </style>
  </defs>`;

  let y = padding;

  if (ast.comments?.length) {
    svg += `<g transform="translate(${padding}, ${y})">`;
    ast.comments.forEach((comment, index) => {
      svg += `<text x="0" y="${15 + index * 20}" class="header-text" style="fill: #666;">${comment.text}</text>`;
    });
    svg += `</g>`;
    y += ast.comments.length * 20 + 20;
    svg += `<line x1="${padding}" y1="${y}" x2="${width - padding}" y2="${y}" class="divider"/>`;
    y += 20;
  }

  if (ast.brewer || ast.dose || ast.temperature) {
    svg += `<g transform="translate(${padding}, ${y})">`;

    const headerItems = [
      ast.brewer && { icon: "brewer", text: ast.brewer },
      ast.dose && { icon: "coffee", text: `${ast.dose}g` },
      ast.temperature && {
        icon: "thermometer",
        text:
          typeof ast.temperature === "number"
            ? `${ast.temperature}째C`
            : `${ast.temperature.min}-${ast.temperature.max}째C`,
      },
    ].filter(Boolean);

    let xOffset = 0;
    headerItems.forEach((item) => {
      if (item) {
        svg += `<g transform="translate(${xOffset}, 0)">
          <g class="icon">${icons[item.icon as keyof typeof icons]}</g>
          <text x="30" y="15" class="header-text">${item.text}</text>
        </g>`;
        xOffset += 100;
      }
    });

    svg += `</g>`;

    y += 40;
    svg += `<line x1="${padding}" y1="${y}" x2="${width - padding}" y2="${y}" class="divider"/>`;
    y += 40;
  }

  let currentTime: Duration = { minutes: 0, seconds: 0 };

  const totalDuration = ast.steps.reduce(
    (sum, step) => addDuration(sum, step.time),
    { minutes: 0, seconds: 0 },
  );

  const ratio = calculateRatioFromAST(ast);

  if (ratio || totalDuration.minutes > 0 || totalDuration.seconds > 0) {
    svg += `<g transform="translate(${padding}, ${y})">`;

    let xOffset = 0;

    if (totalDuration.minutes > 0 || totalDuration.seconds > 0) {
      svg += `<g transform="translate(${xOffset}, 0)">`;
      svg += `<g class="icon">${icons.timer}</g>`;
      svg += `<text x="30" y="15" class="header-text" style="font-weight: 600;">${formatDuration(totalDuration)}</text>`;
      svg += `</g>`;
      xOffset += 150;
    }

    if (ratio.ratio && ratio.water) {
      svg += `<g transform="translate(${xOffset}, 0)">
          <text x="0" y="15" class="header-text" style="font-weight: 600;">${ratio.ratio}</text>
        </g>`;
      xOffset += 120;
      svg += `<g transform="translate(${xOffset}, 0)">
          <g class="icon">${icons.water}</g>
          <text x="30" y="15" class="header-text" style="font-weight: 600;">${ratio.water}g</text>
        </g>`;
    }

    svg += `</g>`;
    y += 40;
    svg += `<line x1="${padding}" y1="${y}" x2="${width - padding}" y2="${y}" class="divider"/>`;
    y += 40;
  }

  ast.steps.forEach((step, index) => {
    const stepEndTime = addDuration(currentTime, step.time);
    const timeRange = formatTimeRange(currentTime, stepEndTime);

    svg += `<g transform="translate(${padding}, ${y})">`;

    svg += `<text x="0" y="20" class="step-number">${index + 1}.</text>`;

    svg += `<g class="icon" transform="translate(30, 4)">${icons.timer}</g>`;
    svg += `<text x="60" y="20" class="time-text">${timeRange}</text>`;

    const stepComment = step.comments?.[0]?.text;
    if (stepComment) {
      svg += `<text x="200" y="20" class="time-text" style="font-style: italic; fill: #888;">${stepComment}</text>`;
    }

    let instructionX = 30;
    const instructionY = 45;

    step.instructions.forEach((instruction) => {
      svg += `<g transform="translate(${instructionX}, ${instructionY})">`;

      switch (instruction.type) {
        case "pour":
          svg += `<g class="icon" transform="translate(0, -12)">${icons.water}</g>`;
          svg += `<text x="30" y="0" class="instruction-text">${instruction.value}g</text>`;
          instructionX += 80;
          break;
        case "swirl":
          svg += `<g class="icon" transform="translate(0, -12)">${icons.swirl}</g>`;
          svg += `<text x="30" y="0" class="instruction-text">swirl</text>`;
          instructionX += 40;
          break;
        case "stir":
          svg += `<g class="icon" transform="translate(0, -12)">${icons.stir}</g>`;
          svg += `<text x="30" y="0" class="instruction-text">stir</text>`;
          instructionX += 40;
          break;
      }

      svg += `</g>`;
    });

    if (step.temperature) {
      const tempText =
        typeof step.temperature === "number"
          ? `${step.temperature}째C`
          : `${step.temperature.min}-${step.temperature.max}째C`;
      svg += `<g transform="translate(${contentWidth - 100}, 30)">
        <g class="icon" transform="translate(0, -12)">${icons.thermometer}</g>
        <text x="30" y="0" class="instruction-text">${tempText}</text>
      </g>`;
    }

    svg += `</g>`;

    currentTime = stepEndTime;
    y += stepHeight;

    if (index < totalSteps - 1) {
      svg += `<line x1="${padding + 30}" y1="${y - 20}" x2="${width - padding}" y2="${y - 20}" class="divider"/>`;
    }
  });

  svg += `</svg>`;

  return svg;
}
