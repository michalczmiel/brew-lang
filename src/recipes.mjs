export const glitchCoffeeOrigamiHot = `
temperature 86..90
dose 14.5
water 260

step {
  start 0:00
  finish 0:05
  pour 10
}

step {
  start 0:30
  finish 0:45
  pour 100
}

step {
  start 1:10
  finish 1:25
  pour 100
}
`;
