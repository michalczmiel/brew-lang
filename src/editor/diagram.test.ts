import { test, expect } from "bun:test";

import grammar from "../core/grammar.ohm-bundle.js";
import { newSemantics, type RecipeAST } from "../core/semantics.js";
import { recipes } from "../recipes.js";
import { generateSVGDiagram } from "./diagram.js";

function getAST(content: string): RecipeAST {
  const semantics = newSemantics(grammar);
  const match = grammar.match(content);

  return semantics(match).toAST();
}

test.each(Object.entries(recipes))(
  "generateSVGDiagram with %s",
  async (recipe, content) => {
    const ast = getAST(content);

    const svg = generateSVGDiagram(ast);
    await Bun.write(`testdata/${recipe}.svg`, svg);

    expect(svg).toMatchSnapshot(recipe);
  },
);
