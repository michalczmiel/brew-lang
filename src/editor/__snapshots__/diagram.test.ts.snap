// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`generateSVGDiagram with glitchCoffeeOrigamiHot: glitchCoffeeOrigamiHot 1`] = `
"<svg viewBox="0 0 600 580" xmlns="http://www.w3.org/2000/svg" style="font-family: system-ui, -apple-system, sans-serif; background: white;"><defs>
    <style>
      .header-text { font-size: 14px; fill: #666; }
      .step-number { font-size: 16px; font-weight: 600; fill: #333; }
      .time-text { font-size: 13px; fill: #666; }
      .instruction-text { font-size: 12px; fill: #444; }
      .icon { color: #333; }
      .divider { stroke: #e0e0e0; stroke-width: 1; stroke-dasharray: 3,3; }
    </style>
  </defs><g transform="translate(20, 20)"><text x="0" y="15" class="header-text" style="fill: #666;">Glitch Coffee Origami Hot Recipe</text></g><line x1="20" y1="60" x2="580" y2="60" class="divider"/><g transform="translate(20, 80)"><g transform="translate(0, 0)">
          <g class="icon"><path d="M8 4h8v3c0 4-2 7-4 7s-4-3-4-7V4z" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M6 4h12" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/><path d="M10 14v4h4v-4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g>
          <text x="30" y="15" class="header-text">origami</text>
        </g><g transform="translate(100, 0)">
          <g class="icon"><circle cx="12" cy="12" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="12" r="1" fill="currentColor"/></g>
          <text x="30" y="15" class="header-text">14.5g</text>
        </g><g transform="translate(200, 0)">
          <g class="icon"><rect x="10" y="4" width="4" height="12" rx="2" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="1.5" fill="currentColor"/><line x1="12" y1="8" x2="12" y2="16" stroke="currentColor" stroke-width="1.5"/></g>
          <text x="30" y="15" class="header-text">86-90°C</text>
        </g></g><line x1="20" y1="120" x2="580" y2="120" class="divider"/><g transform="translate(20, 160)"><g transform="translate(0, 0)"><g class="icon"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="30" y="15" class="header-text" style="font-weight: 600;">2:30</text></g><g transform="translate(150, 0)">
          <text x="0" y="15" class="header-text" style="font-weight: 600;">1:17.9</text>
        </g><g transform="translate(270, 0)">
          <g class="icon"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g>
          <text x="30" y="15" class="header-text" style="font-weight: 600;">260g</text>
        </g></g><line x1="20" y1="200" x2="580" y2="200" class="divider"/><g transform="translate(20, 240)"><text x="0" y="20" class="step-number">1.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:00 - 0:00</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">30g</text></g><g transform="translate(110, 45)"><text x="0" y="0" class="instruction-text">wait 0:05</text></g></g><line x1="50" y1="300" x2="580" y2="300" class="divider"/><g transform="translate(20, 320)"><text x="0" y="20" class="step-number">2.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:00 - 0:20</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">30g</text></g><g transform="translate(110, 45)"><text x="0" y="0" class="instruction-text">wait 0:05</text></g></g><line x1="50" y1="380" x2="580" y2="380" class="divider"/><g transform="translate(20, 400)"><text x="0" y="20" class="step-number">3.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:20 - 1:10</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">100g</text></g><g transform="translate(110, 45)"><text x="0" y="0" class="instruction-text">wait 0:15</text></g></g><line x1="50" y1="460" x2="580" y2="460" class="divider"/><g transform="translate(20, 480)"><text x="0" y="20" class="step-number">4.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">1:10 - 2:30</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">100g</text></g><g transform="translate(110, 45)"><text x="0" y="0" class="instruction-text">wait 0:15</text></g></g></svg>"
`;

exports[`generateSVGDiagram with jamesHoffmannAeropress: jamesHoffmannAeropress 1`] = `
"<svg viewBox="0 0 600 500" xmlns="http://www.w3.org/2000/svg" style="font-family: system-ui, -apple-system, sans-serif; background: white;"><defs>
    <style>
      .header-text { font-size: 14px; fill: #666; }
      .step-number { font-size: 16px; font-weight: 600; fill: #333; }
      .time-text { font-size: 13px; fill: #666; }
      .instruction-text { font-size: 12px; fill: #444; }
      .icon { color: #333; }
      .divider { stroke: #e0e0e0; stroke-width: 1; stroke-dasharray: 3,3; }
    </style>
  </defs><g transform="translate(20, 20)"><text x="0" y="15" class="header-text" style="fill: #666;">James Hoffmann Ultimate AeroPress</text></g><line x1="20" y1="60" x2="580" y2="60" class="divider"/><g transform="translate(20, 80)"><g transform="translate(0, 0)">
          <g class="icon"><path d="M8 4h8v3c0 4-2 7-4 7s-4-3-4-7V4z" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M6 4h12" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/><path d="M10 14v4h4v-4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g>
          <text x="30" y="15" class="header-text">aeropress</text>
        </g><g transform="translate(100, 0)">
          <g class="icon"><circle cx="12" cy="12" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="12" r="1" fill="currentColor"/></g>
          <text x="30" y="15" class="header-text">11g</text>
        </g><g transform="translate(200, 0)">
          <g class="icon"><rect x="10" y="4" width="4" height="12" rx="2" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="1.5" fill="currentColor"/><line x1="12" y1="8" x2="12" y2="16" stroke="currentColor" stroke-width="1.5"/></g>
          <text x="30" y="15" class="header-text">85-99°C</text>
        </g></g><line x1="20" y1="120" x2="580" y2="120" class="divider"/><g transform="translate(20, 160)"><g transform="translate(0, 0)"><g class="icon"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="30" y="15" class="header-text" style="font-weight: 600;">4:30</text></g><g transform="translate(150, 0)">
          <text x="0" y="15" class="header-text" style="font-weight: 600;">1:18.2</text>
        </g><g transform="translate(270, 0)">
          <g class="icon"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g>
          <text x="30" y="15" class="header-text" style="font-weight: 600;">200g</text>
        </g></g><line x1="20" y1="200" x2="580" y2="200" class="divider"/><g transform="translate(20, 240)"><text x="0" y="20" class="step-number">1.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:00 - 0:00</text><text x="200" y="20" class="time-text" style="font-style: italic; fill: #888;">don't rinse paper</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">200g</text></g></g><line x1="50" y1="300" x2="580" y2="300" class="divider"/><g transform="translate(20, 320)"><text x="0" y="20" class="step-number">2.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:00 - 2:00</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 6c0 0 3 2 3 6s-3 6-3 6" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/><path d="M12 6c0 0 -3 2 -3 6s3 6 3 6" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="30" y="0" class="instruction-text">swirl</text></g></g><line x1="50" y1="380" x2="580" y2="380" class="divider"/><g transform="translate(20, 400)"><text x="0" y="20" class="step-number">3.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">2:00 - 4:30</text><text x="200" y="20" class="time-text" style="font-style: italic; fill: #888;">press</text></g></svg>"
`;

exports[`generateSVGDiagram with tetsuKasuyaHybridMethod: tetsuKasuyaHybridMethod 1`] = `
"<svg viewBox="0 0 600 580" xmlns="http://www.w3.org/2000/svg" style="font-family: system-ui, -apple-system, sans-serif; background: white;"><defs>
    <style>
      .header-text { font-size: 14px; fill: #666; }
      .step-number { font-size: 16px; font-weight: 600; fill: #333; }
      .time-text { font-size: 13px; fill: #666; }
      .instruction-text { font-size: 12px; fill: #444; }
      .icon { color: #333; }
      .divider { stroke: #e0e0e0; stroke-width: 1; stroke-dasharray: 3,3; }
    </style>
  </defs><g transform="translate(20, 20)"><text x="0" y="15" class="header-text" style="fill: #666;">Tetsu Kasuya Hybrid method</text></g><line x1="20" y1="60" x2="580" y2="60" class="divider"/><g transform="translate(20, 80)"><g transform="translate(0, 0)">
          <g class="icon"><path d="M8 4h8v3c0 4-2 7-4 7s-4-3-4-7V4z" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M6 4h12" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/><path d="M10 14v4h4v-4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g>
          <text x="30" y="15" class="header-text">hario switch</text>
        </g><g transform="translate(100, 0)">
          <g class="icon"><circle cx="12" cy="12" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="12" r="1" fill="currentColor"/></g>
          <text x="30" y="15" class="header-text">20g</text>
        </g></g><line x1="20" y1="120" x2="580" y2="120" class="divider"/><g transform="translate(20, 160)"><g transform="translate(0, 0)"><g class="icon"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="30" y="15" class="header-text" style="font-weight: 600;">3:30</text></g><g transform="translate(150, 0)">
          <text x="0" y="15" class="header-text" style="font-weight: 600;">1:14.0</text>
        </g><g transform="translate(270, 0)">
          <g class="icon"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g>
          <text x="30" y="15" class="header-text" style="font-weight: 600;">280g</text>
        </g></g><line x1="20" y1="200" x2="580" y2="200" class="divider"/><g transform="translate(20, 240)"><text x="0" y="20" class="step-number">1.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:00 - 0:00</text><text x="200" y="20" class="time-text" style="font-style: italic; fill: #888;">open switch</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">60g</text></g><g transform="translate(460, 30)">
        <g class="icon" transform="translate(0, -12)"><rect x="10" y="4" width="4" height="12" rx="2" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="1.5" fill="currentColor"/><line x1="12" y1="8" x2="12" y2="16" stroke="currentColor" stroke-width="1.5"/></g>
        <text x="30" y="0" class="instruction-text">90°C</text>
      </g></g><line x1="50" y1="300" x2="580" y2="300" class="divider"/><g transform="translate(20, 320)"><text x="0" y="20" class="step-number">2.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:00 - 0:30</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">60g</text></g><g transform="translate(460, 30)">
        <g class="icon" transform="translate(0, -12)"><rect x="10" y="4" width="4" height="12" rx="2" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="1.5" fill="currentColor"/><line x1="12" y1="8" x2="12" y2="16" stroke="currentColor" stroke-width="1.5"/></g>
        <text x="30" y="0" class="instruction-text">90°C</text>
      </g></g><line x1="50" y1="380" x2="580" y2="380" class="divider"/><g transform="translate(20, 400)"><text x="0" y="20" class="step-number">3.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">0:30 - 1:45</text><text x="200" y="20" class="time-text" style="font-style: italic; fill: #888;">close switch</text><g transform="translate(30, 45)"><g class="icon" transform="translate(0, -12)"><path d="M12 2.5s-4 4.5-4 8c0 2.21 1.79 4 4 4s4-1.79 4-4c0-3.5-4-8-4-8z" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/></g><text x="30" y="0" class="instruction-text">160g</text></g><g transform="translate(460, 30)">
        <g class="icon" transform="translate(0, -12)"><rect x="10" y="4" width="4" height="12" rx="2" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="3" fill="none" stroke="currentColor" stroke-width="1.5"/><circle cx="12" cy="18" r="1.5" fill="currentColor"/><line x1="12" y1="8" x2="12" y2="16" stroke="currentColor" stroke-width="1.5"/></g>
        <text x="30" y="0" class="instruction-text">70°C</text>
      </g></g><line x1="50" y1="460" x2="580" y2="460" class="divider"/><g transform="translate(20, 480)"><text x="0" y="20" class="step-number">4.</text><g class="icon" transform="translate(30, 4)"><circle cx="12" cy="13" r="7" fill="none" stroke="currentColor" stroke-width="1.5"/><path d="M12 9v4l2 2" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/><path d="M10 2h4" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></g><text x="60" y="20" class="time-text">1:45 - 3:30</text><text x="200" y="20" class="time-text" style="font-style: italic; fill: #888;">open switch</text></g></svg>"
`;
